File: Section6_Mapping_with_known_Poses/bumperbot_ws/src/bumperbot_utils/src/safety_stop.cpp
#include <string>
#include <memory>
#include <chrono>
#include <thread>

#include "rclcpp/rclcpp.hpp"
#include "sensor_msgs/msg/laser_scan.hpp"
#include "std_msgs/msg/bool.hpp"
#include "twist_mux_msgs/action/joy_turbo.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "visualization_msgs/msg/marker_array.hpp"

using namespace std::chrono_literals;
using std::placeholders::_1;

enum State{
  FREE = 0,
  WARNING,
  DANGER
};

class SafetyStop : public rclcpp::Node
{
public:
  SafetyStop() : Node("safety_stop_node"), is_first_msg_(true), state_{State::FREE}, prev_state_{State::FREE}
  {
    declare_parameter<double>("warning_distance", 0.6);
    declare_parameter<double>("danger_distance", 0.2);
    declare_parameter<std::string>("scan_topic", "scan");
    declare_parameter<std::string>("safety_stop_topic", "safety_stop");
    warning_distance_ = get_parameter("warning_distance").as_double();
    danger_distance_ = get_parameter("danger_distance").as_double();
    std::string scan_topic = get_parameter("scan_topic").as_string();
    std::string safety_stop_topic = get_parameter("safety_stop_topic").as_string();

    laser_sub_ = create_subscription<sensor_msgs::msg::LaserScan>(
        scan_topic, 10, std::bind(&SafetyStop::laserCallback, this, _1));
    safety_stop_pub_ = create_publisher<std_msgs::msg::Bool>(safety_stop_topic, 10);
    zones_pub_ = create_publisher<visualization_msgs::msg::MarkerArray>("zones", 10);
    decrease_speed_client_ = rclcpp_action::create_client<twist_mux_msgs::action::JoyTurbo>(this, "joy_turbo_decrease");
    increase_speed_client_ = rclcpp_action::create_client<twist_mux_msgs::action::JoyTurbo>(this, "joy_turbo_increase");
    while(!decrease_speed_client_->wait_for_action_server(1s) && rclcpp::ok())
    {
      RCLCPP_WARN(get_logger(), "Action /joy_turbo_decrease not available! Waiting..");
      std::this_thread::sleep_for(2s);
    }
    while(!increase_speed_client_->wait_for_action_server(1s) && rclcpp::ok())
    {
      RCLCPP_WARN(get_logger(), "Action /joy_turbo_increase not available! Waiting..");
      std::this_thread::sleep_for(2s);
    }

    // Prepare Zones for Visualization
    visualization_msgs::msg::Marker warning_zone;
    warning_zone.id = 0;
    warning_zone.type = visualization_msgs::msg::Marker::CYLINDER;
    warning_zone.action = visualization_msgs::msg::Marker::ADD;
    warning_zone.scale.z = 0.001;
    warning_zone.scale.x = warning_distance_ * 2;
    warning_zone.scale.y = warning_distance_ * 2;
    warning_zone.color.r = 1.0;
    warning_zone.color.g = 0.984;
    warning_zone.color.b = 0.0;
    warning_zone.color.a = 0.5;
    zones_.markers.push_back(warning_zone);
    visualization_msgs::msg::Marker danger_zone = warning_zone;
    danger_zone.id = 1;
    danger_zone.scale.x = danger_distance_ * 2;
    danger_zone.scale.y = danger_distance_ * 2;
    danger_zone.color.r = 1.0;
    danger_zone.color.g = 0.0;
    danger_zone.color.b = 0.0;
    danger_zone.color.a = 0.5;
    danger_zone.pose.position.z = 0.001;
    zones_.markers.push_back(danger_zone);
  }

private:
  double warning_distance_, danger_distance_;
  bool is_first_msg_;
  State state_, prev_state_;
  visualization_msgs::msg::MarkerArray zones_;
  rclcpp::Subscription<sensor_msgs::msg::LaserScan>::SharedPtr laser_sub_;
  rclcpp::Publisher<std_msgs::msg::Bool>::SharedPtr safety_stop_pub_;
  rclcpp::Publisher<visualization_msgs::msg::MarkerArray>::SharedPtr zones_pub_;
  rclcpp_action::Client<twist_mux_msgs::action::JoyTurbo>::SharedPtr decrease_speed_client_;
  rclcpp_action::Client<twist_mux_msgs::action::JoyTurbo>::SharedPtr increase_speed_client_;

  void laserCallback(const sensor_msgs::msg::LaserScan &msg)
  {
    state_ = State::FREE;
    for (const auto &range : msg.ranges)
    {
      if (range <= warning_distance_)
      {
        state_ = State::WARNING;
        
        if (range <= danger_distance_)
        {
          state_ = State::DANGER;
          // Stop immediately!
          break;
        }
      }
    }

    if(state_ != prev_state_)
    {
      std_msgs::msg::Bool is_safety_stop;
      if(state_ == State::WARNING)
      {
        is_safety_stop.data = false;
        decrease_speed_client_->async_send_goal(twist_mux_msgs::action::JoyTurbo::Goal());
        zones_.markers.at(0).color.a = 1.0;
        zones_.markers.at(1).color.a = 0.5;
      }
      else if(state_ == State::DANGER)
      {
        is_safety_stop.data = true;
        zones_.markers.at(0).color.a = 1.0;
        zones_.markers.at(1).color.a = 1.0;
      }
      else if(state_ == State::FREE)
      {
        is_safety_stop.data = false;
        increase_speed_client_->async_send_goal(twist_mux_msgs::action::JoyTurbo::Goal());
        zones_.markers.at(0).color.a = 0.5;
        zones_.markers.at(1).color.a = 0.5;
      }

      prev_state_ = state_;
      safety_stop_pub_->publish(is_safety_stop);
    }

    
    if(is_first_msg_)
    {
      for(auto & zone : zones_.markers)
      {
        zone.header.frame_id = msg.header.frame_id;
      }
      is_first_msg_ = false;
    }
    zones_pub_->publish(zones_);
  }
};

int main(int argc, char *argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<SafetyStop>());
  rclcpp::shutdown();
  return 0;
}

File: Section8_SLAM/bumperbot_ws/src/bumperbot_utils/bumperbot_utils/safety_stop.py
#!/usr/bin/env python3
import time
from enum import Enum

import rclpy
from rclpy.node import Node
from sensor_msgs.msg import LaserScan
from std_msgs.msg import Bool
from twist_mux_msgs.action import JoyTurbo
from rclpy.action import ActionClient
from visualization_msgs.msg import Marker, MarkerArray


class State(Enum):
    FREE = 0
    WARNING = 1
    DANGER = 2


class SafetyStop(Node):
    def __init__(self):
        super().__init__('safety_stop_node')
        self.declare_parameter('warning_distance', 0.6)
        self.declare_parameter('danger_distance', 0.2)
        self.declare_parameter('scan_topic', 'scan')
        self.declare_parameter('safety_stop_topic', 'safety_stop')
        self.warning_distance = self.get_parameter('warning_distance').get_parameter_value().double_value
        self.danger_distance = self.get_parameter('danger_distance').get_parameter_value().double_value
        self.scan_topic = self.get_parameter('scan_topic').get_parameter_value().string_value
        self.safety_stop_topic = self.get_parameter('safety_stop_topic').get_parameter_value().string_value
        self.is_first_msg = True
        self.state = State.FREE
        self.prev_state = State.FREE

        self.laser_sub = self.create_subscription(
            LaserScan, self.scan_topic, self.laser_callback, 10
        )
        self.safety_stop_pub = self.create_publisher(
            Bool, self.safety_stop_topic, 10
        )
        self.zones_pub = self.create_publisher(
            MarkerArray, 'zones', 10
        )
        self.decrease_speed_client = ActionClient(self, JoyTurbo, 'joy_turbo_decrease')
        self.increase_speed_client = ActionClient(self, JoyTurbo, 'joy_turbo_increase')
        while not self.decrease_speed_client.wait_for_server(timeout_sec=1.0) and rclpy.ok():
            self.get_logger().warn('Action /joy_turbo_decrease not available! Waiting..')
            time.sleep(2.0)

        while not self.increase_speed_client.wait_for_server(timeout_sec=1.0) and rclpy.ok():
            self.get_logger().warn('Action /joy_turbo_increase not available! Waiting..')
            time.sleep(2.0)

        # Prepare Zones for Visualization
        self.zones = MarkerArray()
        warning_zone = Marker()
        warning_zone.id = 0
        warning_zone.type = Marker.CYLINDER
        warning_zone.action = Marker.ADD
        warning_zone.scale.z = 0.001
        warning_zone.scale.x = self.warning_distance * 2
        warning_zone.scale.y = self.warning_distance * 2
        warning_zone.color.r = 1.0
        warning_zone.color.g = 0.984
        warning_zone.color.b = 0.0
        warning_zone.color.a = 0.5
        danger_zone = Marker()
        danger_zone.id = 1
        danger_zone.type = Marker.CYLINDER
        danger_zone.action = Marker.ADD
        danger_zone.scale.z = 0.001
        danger_zone.scale.x = self.danger_distance * 2
        danger_zone.scale.y = self.danger_distance * 2
        danger_zone.color.r = 1.0
        danger_zone.color.g = 0.0
        danger_zone.color.b = 0.0
        danger_zone.color.a = 0.5
        danger_zone.pose.position.z = 0.01
        
        self.zones.markers = [warning_zone, danger_zone]


    def laser_callback(self, msg: LaserScan):
        self.state = State.FREE

        for range_value in msg.ranges:

            if range_value <= self.warning_distance:
                self.state = State.WARNING

                if range_value <= self.danger_distance:
                    self.state = State.DANGER
                    # Stop immediately!
                    break

        if self.state != self.prev_state:
            is_safety_stop = Bool()

            if self.state == State.WARNING:
                is_safety_stop.data = False
                self.decrease_speed_client.send_goal_async(JoyTurbo.Goal())
                self.zones.markers[0].color.a = 1.0
                self.zones.markers[1].color.a = 0.5
            elif self.state == State.DANGER:
                is_safety_stop.data = True
                self.zones.markers[0].color.a = 1.0
                self.zones.markers[1].color.a = 1.0
            elif self.state == State.FREE:
                is_safety_stop.data = False
                self.increase_speed_client.send_goal_async(JoyTurbo.Goal())

                self.zones.markers[0].color.a = 0.5
                self.zones.markers[1].color.a = 0.5

            self.prev_state = self.state
            self.safety_stop_pub.publish(is_safety_stop)

        if self.is_first_msg:
            for zone in self.zones.markers:
                zone.header.frame_id = msg.header.frame_id

            self.is_first_msg = False

        self.zones_pub.publish(self.zones)


def main():
    rclpy.init()
    node = SafetyStop()
    rclpy.spin(node)
    rclpy.shutdown()


if __name__ == '__main__':
    main()

 
